Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -163,24 +205,54 @@ fileprivate typealias HTTPHeadersStorage = [String: [(String, String)]] // [lowe
 /// can be represented appropriately.
 public struct HTTPHeaders: CustomStringConvertible {
 
-    // [lowerCasedName: [(originalCaseName, value)]
-    private var storage: HTTPHeadersStorage = HTTPHeadersStorage()
-    public var description: String { return storage.description }
-
-    // This is expressly *not* public because it doesn't do anything sensible:
-    // it doesn't return the number of header fields in the structure, just the number
-    // of unique header names. That's not really a useful question to ask, but we need it
-    // in NIOHTTP1 so we're adding it internally. At some point this type should be made
-    // to conform to Collection, and when that's done we can add something a bit more sensible.
-    var count: Int {
-        return storage.count
+    // Because we use CoW implementations HTTPHeaders is also CoW
+    private var buffer: ByteBuffer
+    private var headers: [HTTPHeader]
+    private var continuous: Bool = true
+
+    /// Returns the `String` for the given `HTTPHeaderIndex`.
+    ///
+    /// - parameters:
+    ///     - idx: The index into the underlying storage.
+    /// - returns: The value.
+    private func string(idx: HTTPHeaderIndex) -> String {
+        return self.buffer.getString(at: idx.start, length: idx.length)!
+    }
+
+    /// Return all names.
+    var names: [HTTPHeaderIndex] {
+        return self.headers.map { $0.name }
+    }
+
+    public var description: String {
+        var str = ""
```

Comment:
Probably better to build this up as an array and then create the string in one shot. You can even allocate the array size, because it's `self.headers.count * 5`.

Reply:
sure I can do this... that said I think it not matters much as it should not be called often.